# Momentum Strategy

### Picking a stock universe - e.g. Dow Jones Industrial Average 

### Performance evaluation: checking the performance over every single stock over the last 12 month
### skip the most recent month => buy winners & short losers 

#### Highest quantile = winners 
#### lowest quintile = losers 

##### Calculate the strategy return and compare return to the benchmark -> Dow Jones Performance

In [5]:
import yfinance as yf
import pandas_datareader.data as reader
import pandas as pd
import datetime as dt
from dateutil.relativedelta import relativedelta

In [11]:
# Get ticker symbols for the stocks contained in DJI

wiki_url = "https://en.wikipedia.org/wiki/Dow_Jones_Industrial_Average"
tables   = pd.read_html(wiki_url)

# find the table with “Symbol” & “Company”
for tbl in tables:
    if {"Symbol","Company"}.issubset(tbl.columns):
        dji_components = tbl
        break

table = dji_components
tickers = table.Symbol.tolist()
tickers

['MMM',
 'AXP',
 'AMGN',
 'AMZN',
 'AAPL',
 'BA',
 'CAT',
 'CVX',
 'CSCO',
 'KO',
 'DIS',
 'GS',
 'HD',
 'HON',
 'IBM',
 'JNJ',
 'JPM',
 'MCD',
 'MRK',
 'MSFT',
 'NKE',
 'NVDA',
 'PG',
 'CRM',
 'SHW',
 'TRV',
 'UNH',
 'VZ',
 'V',
 'WMT']

In [13]:
# Get prices for the DJI componets
start = dt.datetime(2018,1,31)
end = dt.datetime(2020,1,31)

df = yf.download(tickers, start, end, auto_adjust = False)['Adj Close']
df

[*********************100%***********************]  30 of 30 completed


Ticker           AAPL        AMGN       AMZN         AXP          BA  \
Date                                                                   
2018-01-31  39.293301  148.054565  72.544502   89.722427  337.712067   
2018-02-01  39.375439  147.664581  69.500000   90.264008  340.161255   
2018-02-02  37.666935  148.818481  71.497498   87.267235  332.508820   
2018-02-05  36.725849  139.062271  69.500000   83.051903  313.420319   
2018-02-06  38.260681  140.574234  72.141998   85.010635  324.884796   
...               ...         ...        ...         ...         ...   
2020-01-24  76.963326  190.414429  93.082001  125.538094  321.139313   
2020-01-27  74.700211  188.186050  91.417000  121.384766  314.727478   
2020-01-28  76.813431  189.781357  92.662498  123.066521  314.687714   
2020-01-29  78.421303  191.731186  92.900002  122.351082  320.115387   
2020-01-30  78.307678  190.887115  93.533997  123.782005  321.387817   

Ticker             CAT         CRM       CSCO        CVX         DIS  ...  \
Date                                                                  ...   
2018-01-31  137.699844  112.920532  33.149426  90.738747  103.949203  ...   
2018-02-01  137.243011  111.760689  33.277096  90.898003  105.690147  ...   
2018-02-02  133.224884  109.817711  32.662628  85.838043  103.977890  ...   
2018-02-05  127.802437  106.714905  30.946899  81.523727  100.151672  ...   
2018-02-06  132.311264  108.618233  32.056137  84.824646  101.557808  ...   
...                ...         ...        ...        ...         ...  ...   
2020-01-24  125.143089  180.528122  41.298122  87.506561  137.783051  ...   
2020-01-27  120.997787  179.140289  40.131470  86.364334  133.671600  ...   
2020-01-28  121.898178  181.261688  40.385094  86.935448  136.101089  ...   
2020-01-29  121.042374  180.191055  39.776390  86.348694  133.828979  ...   
2020-01-30  120.676888  184.057205  39.937027  87.154526  135.550262  ...   

Ticker            MSFT        NKE      NVDA          PG         SHW  \
Date                                                                  
2018-01-31   87.690552  62.288891  6.077598   70.879700  130.082962   
2018-02-01   86.998322  61.768444  5.946550   70.477455  128.529816   
2018-02-02   84.709396  61.375816  5.773964   69.163940  125.723038   
2018-02-05   81.220612  58.791855  5.283900   66.545151  122.635536   
2018-02-06   84.294075  59.549698  5.577642   67.628761  126.306168   
...                ...        ...       ...         ...         ...   
2020-01-24  157.253754  95.209564  6.234852  109.161293  188.836624   
2020-01-27  154.623962  93.539230  5.978965  109.641075  185.481018   
2020-01-28  157.653961  93.697861  6.172373  109.937660  187.633423   
2020-01-29  160.112244  93.165970  6.111887  109.091545  187.382599   
2020-01-30  164.628601  91.626282  6.118607  109.867874  180.601715   

Ticker             TRV         UNH           V         VZ        WMT  
Date                                                                  
2018-01-31  127.049706  211.059097  117.970703  36.176826  31.400476  
2018-02-01  127.117516  209.668533  119.385605  36.330704  31.082348  
2018-02-02  125.380287  206.691315  114.818001  35.447533  30.776016  
2018-02-05  119.931168  196.119705  110.411766  33.788227  29.482878  
2018-02-06  119.363358  200.719162  113.925331  34.009018  29.721472  
...                ...         ...         ...        ...        ...  
2020-01-24  119.597443  268.995819  197.225662  43.991627  35.187347  
2020-01-27  118.665718  260.985016  194.041199  43.721596  35.645771  
2020-01-28  119.251373  262.620209  195.157181  44.298126  35.873447  
2020-01-29  118.257553  259.891876  197.090988  43.444279  35.654995  
2020-01-30  119.082787  258.128052  200.313934  43.320213  35.867287  

[503 rows x 30 columns]

In [16]:
# calculating the monthly returns
mtl_ret = df.pct_change().resample('M').agg(lambda x:(x+1).prod() - 1)
mtl_ret

Ticker          AAPL      AMGN      AMZN       AXP        BA       CAT  \
Date                                                                     
2018-01-31  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   
2018-02-28  0.068185 -0.004788  0.042429 -0.019014  0.027169 -0.050068   
2018-03-31 -0.058051 -0.072319 -0.043049 -0.043380 -0.094779 -0.046886   
2018-04-30 -0.015020  0.023463  0.082075  0.062615  0.017323 -0.015560   
2018-05-31  0.135124  0.037404  0.040539 -0.004557  0.061029  0.052300   
2018-06-30 -0.009418  0.027669  0.043065 -0.003052 -0.047279 -0.106905   
2018-07-31  0.027984  0.064792  0.045676  0.019156  0.061965  0.066424   
2018-08-31  0.200422  0.023505  0.132364  0.064912 -0.033164 -0.034423   
2018-09-30 -0.008303  0.037436 -0.004824  0.004812  0.084921  0.098236   
2018-10-31 -0.030477 -0.069951 -0.202192 -0.031827 -0.045819 -0.199336   
2018-11-30 -0.181198  0.087734  0.057672  0.092864 -0.018313  0.118282   
2018-12-31 -0.116699 -0.065210 -0.111350 -0.150975 -0.069962 -0.063389   
2019-01-31  0.055154 -0.038835  0.144317  0.081833  0.195721  0.054668   
2019-02-28  0.044776  0.023728 -0.045906  0.049075  0.146648  0.031391   
2019-03-31  0.097026 -0.000526  0.085936  0.014479 -0.133057 -0.013470   
2019-04-30  0.056436 -0.056111  0.081859  0.076350 -0.009779  0.035243   
2019-05-31 -0.124213 -0.062374 -0.078613 -0.021496 -0.090332 -0.140654   
2019-06-30  0.130519  0.105459  0.066792  0.076105  0.065572  0.137551   
2019-07-31  0.076395  0.012481 -0.014179  0.010685 -0.062718 -0.026464   
2019-08-31 -0.016462  0.126048 -0.048474 -0.032162  0.073808 -0.096226   
2019-09-30  0.072962 -0.072428 -0.022733 -0.017363  0.044989  0.061429   
2019-10-31  0.110684  0.102010  0.023475 -0.004662 -0.106605  0.099542   
2019-11-30  0.077554  0.107994  0.013587  0.024215  0.083572  0.050290   
2019-12-31  0.098784  0.027054  0.026122  0.036380 -0.110383  0.020383   
2020-01-31  0.102911 -0.061891  0.012360  0.073835 -0.007552 -0.076926   

Ticker           CRM      CSCO       CVX       DIS  ...      MSFT       NKE  \
Date                                                ...                       
2018-01-31  0.000000  0.000000  0.000000  0.000000  ...  0.000000  0.000000   
2018-02-28  0.020543  0.077997 -0.098252 -0.050704  ... -0.008415 -0.017444   
2018-03-31  0.000430 -0.042207  0.018942 -0.026367  ... -0.026661 -0.005804   
2018-04-30  0.040327  0.040943  0.097071 -0.001095  ...  0.024652  0.029350   
2018-05-31  0.068931 -0.035674  0.002189 -0.008572  ...  0.061467  0.049861   
2018-06-30  0.054666  0.007493  0.017136  0.053684  ... -0.002327  0.112850   
2018-07-31  0.005499 -0.009537 -0.001265  0.092194  ...  0.075753 -0.034764   
2018-08-31  0.113234  0.129581 -0.052857 -0.013561  ...  0.062993  0.071414   
2018-09-30  0.041590  0.018422  0.032247  0.043921  ...  0.018161  0.030657   
2018-10-31 -0.137018 -0.053252 -0.086932 -0.018043  ... -0.066101 -0.114259   
2018-11-30  0.040221  0.046339  0.075705  0.005748  ...  0.042684  0.004038   
2018-12-31 -0.040558 -0.094840 -0.085337 -0.043203  ... -0.084047 -0.013046   
2019-01-31  0.109513  0.099842  0.053865  0.017054  ...  0.028158  0.104397   
2019-02-28  0.076857  0.094735  0.053485  0.011836  ...  0.077358  0.047020   
2019-03-31 -0.032264  0.042882  0.030105 -0.016040  ...  0.052754 -0.015203   
2019-04-30  0.044074  0.042864 -0.025329  0.233631  ...  0.107343  0.042988   
2019-05-31 -0.084306 -0.070062 -0.042394 -0.035993  ... -0.049481 -0.119268   
2019-06-30  0.002114  0.051893  0.093017  0.057558  ...  0.083118  0.088281   
2019-07-31  0.018256  0.018630 -0.010688  0.030476  ...  0.017243  0.024777   
2019-08-31  0.010162 -0.155054 -0.033952 -0.040207  ...  0.015037 -0.015247   
2019-09-30 -0.048888  0.055544  0.007475 -0.050561  ...  0.008487  0.111479   
2019-10-31  0.054231 -0.031171 -0.020742 -0.003070  ...  0.031216 -0.046529   
2019-11-30  0.040897 -0.046306  0.018462  0.166718  ...  0.059462  0.046722   
2019-12-31 -0.001535  0.05

In [17]:
#calculate the return s of the last 11 month
import numpy as np

past_11 = (mtl_ret+1).rolling(11).apply(np.prod) - 1
past_11

Ticker          AAPL      AMGN      AMZN       AXP        BA       CAT  \
Date                                                                     
2018-01-31       NaN       NaN       NaN       NaN       NaN       NaN   
2018-02-28       NaN       NaN       NaN       NaN       NaN       NaN   
2018-03-31       NaN       NaN       NaN       NaN       NaN       NaN   
2018-04-30       NaN       NaN       NaN       NaN       NaN       NaN   
2018-05-31       NaN       NaN       NaN       NaN       NaN       NaN   
2018-06-30       NaN       NaN       NaN       NaN       NaN       NaN   
2018-07-31       NaN       NaN       NaN       NaN       NaN       NaN   
2018-08-31       NaN       NaN       NaN       NaN       NaN       NaN   
2018-09-30       NaN       NaN       NaN       NaN       NaN       NaN   
2018-10-31       NaN       NaN       NaN       NaN       NaN       NaN   
2018-11-30  0.082608  0.152214  0.164919  0.141893 -0.002270 -0.151840   
2018-12-31 -0.043731  0.077077  0.035206 -0.030504 -0.072073 -0.205605   
2019-01-31 -0.055396  0.040230  0.136388  0.069161  0.080194 -0.118018   
2019-02-28  0.047721  0.147929  0.132996  0.172494  0.368287 -0.045583   
2019-03-31  0.166903  0.121022  0.137038  0.119381  0.166027 -0.043556   
2019-04-30  0.086012  0.019969  0.182190  0.210360  0.088212 -0.059060   
2019-05-31 -0.039841 -0.069399  0.044282  0.187968  0.039036 -0.094617   
2019-06-30  0.055929 -0.033858  0.065369  0.254350  0.042565 -0.034231   
2019-07-31 -0.053169 -0.044264 -0.072504  0.190477  0.010696 -0.026270   
2019-08-31 -0.060959  0.037370 -0.113185  0.146671  0.000344 -0.198686   
2019-09-30  0.039228  0.034606  0.086295  0.163800  0.095545  0.062291   
2019-10-31  0.409686  0.048185  0.051172  0.059943 -0.002987  0.044490   
2019-11-30  0.719699  0.242400  0.198959  0.278656  0.161603  0.171263   
2019-12-31  0.790808  0.327567  0.075119  0.224934 -0.135766  0.133187   
2020-01-31  0.890455  0.216537  0.140777  0.253845 -0.251987  0.014180   

Ticker           CRM      CSCO       CVX       DIS  ...      MSFT       NKE  \
Date                                                ...                       
2018-01-31       NaN       NaN       NaN       NaN  ...       NaN       NaN   
2018-02-28       NaN       NaN       NaN       NaN  ...       NaN       NaN   
2018-03-31       NaN       NaN       NaN       NaN  ...       NaN       NaN   
2018-04-30       NaN       NaN       NaN       NaN  ...       NaN       NaN   
2018-05-31       NaN       NaN       NaN       NaN  ...       NaN       NaN   
2018-06-30       NaN       NaN       NaN       NaN  ...       NaN       NaN   
2018-07-31       NaN       NaN       NaN       NaN  ...       NaN       NaN   
2018-08-31       NaN       NaN       NaN       NaN  ...       NaN       NaN   
2018-09-30       NaN       NaN       NaN       NaN  ...       NaN       NaN   
2018-10-31       NaN       NaN       NaN       NaN  ...       NaN       NaN   
2018-11-30  0.253270  0.178616 -0.014529  0.071302  ...  0.187345  0.113583   
2018-12-31  0.202441  0.066836 -0.098626  0.025019  ...  0.087552  0.099055   
2019-01-31  0.307269  0.088455  0.053428  0.098182  ...  0.127664  0.235342   
2019-02-28  0.407137  0.244077  0.089140  0.141272  ...  0.248175  0.300978   
2019-03-31  0.308951  0.246395  0.022658  0.124197  ...  0.282407  0.244669   
2019-04-30  0.278512  0.347905 -0.005422  0.398834  ...  0.337832  0.236521   
2019-05-31  0.110044  0.244146 -0.063632  0.279781  ...  0.274601 -0.021392   
2019-06-30  0.106307  0.321310  0.024763  0.239197  ...  0.283327  0.103358   
2019-07-31  0.011920  0.191527  0.070388  0.294517  ...  0.228095  0.055330   
2019-08-31 -0.018613 -0.011436  0.001743  0.190194  ...  0.224327  0.008327   
2019-09-30  0.081609  0.102165  0.105318  0.150781  ...  0.322111  0.265307   
2019-10-31  0.096175  0.020520  0.006217  0.140692  ...  0.307570  0.201582   
2019-11-30  0.189238  0.075240  0.120405  0.390960  ...  0.512438  0.274347   
2019-12-31  0.070211  0.03

In [23]:
# define formation day
formation = dt.datetime(2019,12,31)

end_measurement = formation - relativedelta(month=1)

In [21]:
formation

datetime.datetime(2019, 12, 31, 0, 0)

In [24]:
end_measurement

datetime.datetime(2019, 1, 31, 0, 0)

In [32]:
#loc functiom to find the end_measurement in that dataframe 
ret_12 = past_11.loc[end_measurement]
ret_12

Ticker
AAPL   -0.055396
AMGN    0.040230
AMZN    0.136388
AXP     0.069161
BA      0.080194
CAT    -0.118018
CRM     0.307269
CSCO    0.088455
CVX     0.053428
DIS     0.098182
GS     -0.238673
HD      0.029667
HON     0.007208
IBM    -0.107167
JNJ     0.045734
JPM    -0.080711
KO      0.152121
MCD     0.154781
MMM    -0.132062
MRK     0.414817
MSFT    0.127664
NKE     0.235342
NVDA   -0.404704
PG      0.272267
SHW     0.056284
TRV    -0.075655
UNH     0.210841
V       0.103574
VZ      0.206939
WMT     0.089929
Name: 2019-01-31 00:00:00, dtype: float64

In [37]:
#transform series into dataframe 
ret_12 = ret_12.reset_index()
ret_12

Ticker  2019-01-31 00:00:00
0    AAPL            -0.055396
1    AMGN             0.040230
2    AMZN             0.136388
3     AXP             0.069161
4      BA             0.080194
5     CAT            -0.118018
6     CRM             0.307269
7    CSCO             0.088455
8     CVX             0.053428
9     DIS             0.098182
10     GS            -0.238673
11     HD             0.029667
12    HON             0.007208
13    IBM            -0.107167
14    JNJ             0.045734
15    JPM            -0.080711
16     KO             0.152121
17    MCD             0.154781
18    MMM            -0.132062
19    MRK             0.414817
20   MSFT             0.127664
21    NKE             0.235342
22   NVDA            -0.404704
23     PG             0.272267
24    SHW             0.056284
25    TRV            -0.075655
26    UNH             0.210841
27      V             0.103574
28     VZ             0.206939
29    WMT             0.089929

In [38]:
# use quintiles to divide the distribution into 5 equal parts to see the winners [4} and losers [0]
# 1) coerce your returns columns to numeric (so numpy.isnan will work)
ret_12.iloc[:,1:] = ret_12.iloc[:,1:].apply(pd.to_numeric, errors='coerce')

In [40]:
# 2a) If you just want **one** quintile column based on a single return series (say the 2nd column):
ret_12['quintile'] = pd.qcut(
    ret_12.iloc[:,1],  # ← note: a Series, not a DataFrame
    5,
    labels=False
)
ret_12

Ticker  2019-01-31 00:00:00  quintile
0    AAPL            -0.055396         1
1    AMGN             0.040230         1
2    AMZN             0.136388         3
3     AXP             0.069161         2
4      BA             0.080194         2
5     CAT            -0.118018         0
6     CRM             0.307269         4
7    CSCO             0.088455         2
8     CVX             0.053428         2
9     DIS             0.098182         3
10     GS            -0.238673         0
11     HD             0.029667         1
12    HON             0.007208         1
13    IBM            -0.107167         0
14    JNJ             0.045734         1
15    JPM            -0.080711         0
16     KO             0.152121         3
17    MCD             0.154781         3
18    MMM            -0.132062         0
19    MRK             0.414817         4
20   MSFT             0.127664         3
21    NKE             0.235342         4
22   NVDA            -0.404704         0
23     PG             0.272267         4
24    SHW             0.056284         2
25    TRV            -0.075655         1
26    UNH             0.210841         4
27      V             0.103574         3
28     VZ             0.206939         4
29    WMT             0.089929         2

In [42]:
winners = ret_12[ret_12.quintile == 4]
losers = ret_12[ret_12.quintile == 0]

In [43]:
winners

Ticker  2019-01-31 00:00:00  quintile
6     CRM             0.307269         4
19    MRK             0.414817         4
21    NKE             0.235342         4
23     PG             0.272267         4
26    UNH             0.210841         4
28     VZ             0.206939         4

In [44]:
losers

Ticker  2019-01-31 00:00:00  quintile
5     CAT            -0.118018         0
10     GS            -0.238673         0
13    IBM            -0.107167         0
15    JPM            -0.080711         0
18    MMM            -0.132062         0
22   NVDA            -0.404704         0

In [53]:
winner_ret = mtl_ret.loc[formation + relativedelta(month=1), df.columns.isin(winners.Ticker)]
loser_ret = mtl_ret.loc[formation + relativedelta(month=1), df.columns.isin(losers.Ticker)]

In [47]:
winner_ret

Ticker
CRM    0.109513
MRK   -0.025913
NKE    0.104397
PG     0.057800
UNH    0.084618
VZ    -0.010412
Name: 2019-01-31 00:00:00, dtype: float64

In [48]:
loser_ret

Ticker
CAT     0.054668
GS      0.185333
IBM     0.182546
JPM     0.068844
MMM     0.051223
NVDA    0.076779
Name: 2019-01-31 00:00:00, dtype: float64

In [57]:
#equal weighting the portfolio
momentum = winner_ret.mean() - loser_ret.mean()

In [58]:
momentum

-0.04989807332815863

In [60]:
#check how DJI performed in Jan to compare
DJI = yf.download('^DJI', start, end, auto_adjust = False)['Adj Close'].pct_change().resample('M').agg(lambda x: (x+1).prod() -1 )
DJI

[*********************100%***********************]  1 of 1 completed


Ticker          ^DJI
Date                
2018-01-31  0.000000
2018-02-28 -0.042838
2018-03-31 -0.037000
2018-04-30  0.002491
2018-05-31  0.010458
2018-06-30 -0.005915
2018-07-31  0.047125
2018-08-31  0.021626
2018-09-30  0.019006
2018-10-31 -0.050742
2018-11-30  0.016830
2018-12-31 -0.086575
2019-01-31  0.071684
2019-02-28  0.036654
2019-03-31  0.000489
2019-04-30  0.025618
2019-05-31 -0.066855
2019-06-30  0.071929
2019-07-31  0.009936
2019-08-31 -0.017160
2019-09-30  0.019450
2019-10-31  0.004807
2019-11-30  0.037165
2019-12-31  0.017362
2020-01-31  0.011248

### One month after formation (i.e. in February 2019), you pulled each group’s month‑end return:

### Winners’ Feb‑2019 returns:
### CRM +10.95%, MRK −2.59%, NKE +10.44%, PG +5.78%, UNH +8.46%, VZ −1.04%

### Losers’ Feb‑2019 returns:
### CAT +5.47%, GS +18.53%, IBM +18.25%, JPM +6.88%, MMM +5.12%, NVDA +7.68% 

## losing about -4.9% per month
## benchmark - + 3.66% per month

### Negative strategy return (–4.99%) means that in this single formation → holding window the “classic” momentum signal failed: the bottom‐past‑11‑month group actually outperformed the top group.

### DJIA benchmark (+3.67%) was positive—so a simple “buy index” would have beaten your long‑winners/short‑losers trade by ~8.7 percentage points that month.